src/imp/linux_raw/syscalls.rs

```diff
@@ -3012,6 +3017,56 @@ pub(crate) fn gettid() -> Pid {
     }
 }
 
+/// https://github.com/ifduyue/musl/blob/cfdfd5ea3ce14c6abf7fb22a531f3d99518b5a1b/src/internal/procfdname.c
+fn procfdname(mut fd: i32) -> [u8; 24] {
+    let mut buf = [0; 24];
+    let mut i = 0;
+
+    for c in "/proc/self/fd/".as_bytes() {
+        buf[i] = *c;
+        i += 1;
+    }
+
+    if fd == 0 {
+        buf[i] = '0' as u8;
+        buf[i + 1] = 0;
+    } else {
+        let mut j = fd;
+        while j != 0 {
+            j /= 10;
+            i += 1;
+        }
+
+        buf[i] = 0;
+
+        while fd != 0 {
+            i -= 1;
+            buf[i] = '0' as u8 + (fd % 10) as u8;
+            fd /= 10;
+        }
+    }
+
+    buf
+}
+
+#[inline]
+pub(crate) fn ttyname(fd: BorrowedFd<'_>, buf: &mut [u8]) -> io::Result<()> {
+    // Check that the fd is really a tty
+    ioctl_tiocgwinsz(fd)?;
+
+    // Get the proc path of the tty
+    let procname = procfdname(fd.as_raw_fd());
+
+    // And finaly gatter the ttyname by reading the proc path
+    readlink(
+        // SATEFY: procfdname always return a nul terminated string
+        unsafe { CStr::from_bytes_with_nul_unchecked(&procname) },
+        buf,
+    )?;
```

Comment:
`readlink("/proc/self/fd/...")` isn't always reliable; file handles for deleted files or pipes end up being symlinks to special strings which aren't paths. We should check that the resulting path can be opened with `stat` and that the result's `st_dev`+`st_ino` matches that of `fd`.

Reply:
Yeah you're right, I was a little bit lazy after 25 github actions to get every `cfg` right. Done.

Reply:
I can sympathize. It's a lot :-}.